### PR TITLE
feat: multicloud upload

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

### Overview

- Allow multicloud upload

### Rationale

- To call the application once without having to parallelize the upload process for each cloud.

### Checklist

- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The documentation is generated using `src-docs`
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)
- [x] The application version in `pyproject.toml` is incremented

### CODE REVIEW

File: src/github_runner_image_builder/cli.py [hunk 1]
```diff
@@ -148,14 +148,14 @@ def get_latest_build_id(cloud_name: str, image_name: str) -> None:
...
-    help="EXPERIMENTAL: Different cloud to use to upload the externally built image. The cloud "
-    "connection parameters should exist in the clouds.yaml. Ignored if --experimental-external is"
-    " not enabled, as a part of external build mode parameter.",
+    help="EXPERIMENTAL: Comma separated list of different clouds to use to upload the externally "
+    "built image. The cloud connection parameters should exist in the clouds.yaml. Ignored if "
+    "--experimental-external is not enabled, as a part of external build mode parameter.",
 )
 # click doesn't yet support dataclasses, hence all arguments are required.
-def run(  # pylint: disable=too-many-arguments
+def run(  # pylint: disable=too-many-arguments, too-many-locals
```
Comment (jdkandersson) [inline src/github_runner_image_builder/cli.py:158]: Perhaps split this into multiple functions? This linting rule being suppressed indicates that the function is becoming too complex

File: src/github_runner_image_builder/openstack_builder.py [hunk 2]
```diff
@@ -521,3 +514,59 @@ def _wait_for_snapshot_complete(
...
+        conn: The OpenStack connection instance.
+        image: The snapshot image to upload.
+        upload_cloud_names: The clouds to upload the image to.
+        upload_cloud_config: The upload image configuration.
+
+    Returns:
+        The uploaded cloud images.
+    """
+    if not upload_cloud_names:
+        return (image,)
```
Comment (jdkandersson) [inline src/github_runner_image_builder/openstack_builder.py:552]: Perhaps just don't call the function in this case?

File: tests/unit/test_openstack_builder.py [hunk 3]
```diff
@@ -184,7 +184,7 @@ def test__create_security_group():
                 flavor="test-flavor",
                 network="test-network",
                 proxy="test-proxy",
-                upload_cloud_name="test-cloud-2",
+                upload_cloud_names=["test-cloud-2"],
```
Comment (jdkandersson) [inline tests/unit/test_openstack_builder.py:187]: Should we also have a case with multiple?

File: src/github_runner_image_builder/openstack_builder.py (hunk 2 above)
Comment (yanksyoon) [reply src/github_runner_image_builder/openstack_builder.py:552]: This is to reduce the if/else statement on the controller function which helps reduce complexity when reading it :)

File: tests/unit/test_openstack_builder.py (hunk 3 above)
Comment (yanksyoon) [reply tests/unit/test_openstack_builder.py]: Added a case for multiple! Thank you for the review :)

File: src/github_runner_image_builder/cli.py (hunk 1 above)
Comment (yanksyoon) [reply src/github_runner_image_builder/cli.py:158]: It isn't complex - it only has 2 if/else statements, the variables come from the number of arguments which cannot be reduced due to the limitation of click library (it doesn't allow for argument wrapper datastructures).

File: src/github_runner_image_builder/openstack_builder.py (hunk 2 above)
Comment (jdkandersson) [reply src/github_runner_image_builder/openstack_builder.py:552]: Can you think of any alternative ways?

Comment (yanksyoon) [reply src/github_runner_image_builder/openstack_builder.py:552]: What would your suggestion be? The caller function acts as a controller and delegates sub-level checks and returns values with consistent types. I think it is fine.

**Review by jdkandersson — APPROVED** [review]

File: src/github_runner_image_builder/openstack_builder.py [hunk 4]
```diff
@@ -197,14 +198,15 @@ class CloudConfig:
         flavor: The OpenStack flavor to launch builder VMs on.
         network: The OpenStack network to launch the builder VMs on.
         proxy: The proxy to enable on builder VMs.
-        upload_cloud_name: The OpenStack cloud name to upload the snapshot to. (Default same cloud)
+        upload_cloud_names: The OpenStack cloud names to upload the snapshot to. (Defaults to \
```
Comment (yhaliaw) [inline src/github_runner_image_builder/openstack_builder.py:201]: Is the `\` line break intended here?

**Review by yhaliaw — APPROVED** [review]

File: src/github_runner_image_builder/openstack_builder.py (hunk 4 above)
Comment (yanksyoon) [reply src/github_runner_image_builder/openstack_builder.py:201]: Yup, the line was too long! Thanks for the review!

**Pull request merged**